crates/rpc/rpc/src/eth/logs_utils.rs

```diff
@@ -1,46 +1,95 @@
 use reth_rpc_types::{FilteredParams, Log};
 use reth_rpc_types_compat::log::from_primitive_log;
 
-/// Returns all matching logs of a block's receipts grouped with the hash of their transaction.
-pub(crate) fn matching_block_logs<'a, I>(
+/// Returns all matching of a block's receipts when the transaction hashes are known.
+pub(crate) fn matching_block_logs_with_tx_hashes<'a, I>(
     filter: &FilteredParams,
-    block: BlockNumHash,
-    tx_and_receipts: I,
+    block_num_hash: BlockNumHash,
+    tx_hashes_and_receipts: I,
     removed: bool,
 ) -> Vec<Log>
 where
     I: IntoIterator<Item = (TxHash, &'a Receipt)>,
 {
     let mut all_logs = Vec::new();
-    append_matching_block_logs(&mut all_logs, filter, block, tx_and_receipts, removed);
+    // Tracks the index of a log in the entire block.
+    let mut log_index: u32 = 0;
+    // Iterate over transaction hashes and receipts and append matching logs.
+    for (receipt_idx, (tx_hash, receipt)) in tx_hashes_and_receipts.into_iter().enumerate() {
+        let logs = &receipt.logs;
+        for log in logs {
+            if log_matches_filter(block_num_hash, log, filter) {
+                let log = Log {
+                    address: log.address,
+                    topics: log.topics.clone(),
+                    data: log.data.clone(),
+                    block_hash: Some(block_num_hash.hash),
+                    block_number: Some(U256::from(block_num_hash.number)),
+                    transaction_hash: Some(tx_hash),
+                    // The transaction and receipt index is always the same.
+                    transaction_index: Some(U256::from(receipt_idx)),
+                    log_index: Some(U256::from(log_index)),
+                    removed,
+                };
+                all_logs.push(log);
+            }
+            log_index += 1;
+        }
+    }
     all_logs
 }
 
-/// Appends all matching logs of a block's receipts grouped with the hash of their transaction
-pub(crate) fn append_matching_block_logs<'a, I>(
+/// Appends all matching logs of a block's receipts.
+/// If the log matches, look up the corresponding transaction hash.
+pub(crate) fn append_matching_block_logs(
     all_logs: &mut Vec<Log>,
+    provider: impl BlockReader,
     filter: &FilteredParams,
-    block: BlockNumHash,
-    tx_and_receipts: I,
+    block_num_hash: BlockNumHash,
+    receipts: &[Receipt],
     removed: bool,
-) where
-    I: IntoIterator<Item = (TxHash, &'a Receipt)>,
-{
-    let block_number_u256 = U256::from(block.number);
-    // tracks the index of a log in the entire block
+) -> Result<(), FilterError> {
+    // Tracks the index of a log in the entire block.
     let mut log_index: u32 = 0;
-    for (transaction_idx, (transaction_hash, receipt)) in tx_and_receipts.into_iter().enumerate() {
-        for log in receipt.logs.iter() {
-            if log_matches_filter(block, log, filter) {
+
+    // Lazy loaded number of the first transaction in the block.
```

Comment:
```suggestion
    // Lazy loaded number of the first transaction in the block.
    // This is useful for blocks with multiple matching logs because it prevents
    // re-querying the block body indices.
```